app/src/crypto.c

```diff
@@ -733,19 +733,23 @@ zxerr_t crypto_extract_spend_signature(uint8_t *buffer, uint16_t bufferLen, uint
     return get_next_spend_signature(buffer);
 }
 
-parser_error_t checkSpends(const parser_tx_t *txObj, keys_t *keys, parser_context_t *builder_spends_ctx, parser_context_t *tx_spends_ctx) {
+parser_error_t checkSpends(const parser_tx_t *txObj, keys_t *keys, parser_context_t *builder_spends_ctx, parser_context_t *tx_spends_ctx, parser_context_t *indices_ctx) {
     if (txObj == NULL || keys == NULL) {
         return parser_unexpected_error;
     }
 
-    if (txObj->transaction.sections.maspBuilder.builder.sapling_builder.n_spends != txObj->transaction.sections.maspTx.data.sapling_bundle.n_shielded_spends) {
+    if (txObj->transaction.sections.maspBuilder.metadata.n_spends_indices != txObj->transaction.sections.maspTx.data.sapling_bundle.n_shielded_spends) {
         return parser_invalid_number_of_spends;
     }
 
-    for (uint32_t i = 0; i < txObj->transaction.sections.maspBuilder.builder.sapling_builder.n_spends; i++) {
+    for (uint32_t i = 0; i < txObj->transaction.sections.maspBuilder.metadata.n_spends_indices; i++) {
         CHECK_ERROR(getNextSpendDescription(builder_spends_ctx, i));
-        CTX_CHECK_AND_ADVANCE(tx_spends_ctx, SHIELDED_SPENDS_LEN * i);
-        spend_item_t *item = spendlist_retrieve_rand_item(i);
+
+        uint64_t indice = 0;
+        CHECK_ERROR(readUint64(indices_ctx, &indice));
+
+        CTX_CHECK_AND_ADVANCE(tx_spends_ctx, SHIELDED_SPENDS_LEN * indice);
+        spend_item_t *item = spendlist_retrieve_rand_item(indice);
```

Comment:
We were previously doing SHIELDAD_SPENDS_LEN * i, and now it is SHIELD_SPENDS_LEN * indice. Is that correct? I assume there is some correlation between i and indice, right?

Reply:
Yes that is correct. The i will go through all the spends in the tx, and the indice, through an array of the actual spends index that should be checked. I just notice this might be wrong, so I will redo the logic, the one used in the outputs that is the correct one.